File: app/schemas/vision.py
from typing import List, Literal

from pydantic import BaseModel, Field


ReviewStatus = Literal["pending", "accepted", "rejected"]


class BoundingBox(BaseModel):

    x: int = Field(..., ge=0, description="Left coordinate of the bounding box in pixels.")
    y: int = Field(..., ge=0, description="Top coordinate of the bounding box in pixels.")
    width: int = Field(..., gt=0, description="Width of the bounding box in pixels.")
    height: int = Field(..., gt=0, description="Height of the bounding box in pixels.")


class VisualDetection(BaseModel):

    label: str = Field(
        default="visual_defect_candidate", description="Human-readable label for the detected visual region.",)
    confidence: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Heuristic confidence score between 0.0 and 1.0.",
    )
    bbox: BoundingBox = Field(..., description="Bounding box of the detected region.")
    review_status: ReviewStatus = Field(
        default="pending",
        description="Human review status for the detection.",
    )


class VisionInspectionResponse(BaseModel):

    filename: str = Field(..., description="Original uploaded filename.")
    image_width: int = Field(..., gt=0, description="Width of the image coordinate space.")
    image_height: int = Field(..., gt=0, description="Height of the image coordinate space.")
    detections: List[VisualDetection] = Field(
        default_factory=list,
        description="List of detected visual regions.",
    )

File: frontend/src/types/vision.ts
export type VisionReviewStatus = "pending" | "accepted" | "rejected"

export type VisionBoundingBox = {
  x: number
  y: number
  width: number
  height: number
}

export type VisionDetection = {
  label: string
  confidence: number
  bbox: VisionBoundingBox
  review_status: VisionReviewStatus
}

export type VisionDetectionWithId = VisionDetection & {
  id: number
}

export type VisionInspectionResponse = {
  inspection_id: number
  filename: string
  image_width: number
  image_height: number
  detections: VisionDetectionWithId[]
}

export type VisionOcrResponse = {
  filename: string
  image_width: number
  image_height: number
  extracted_text: string
  raw_text: string
  char_count: number
  is_empty: boolean
  engine: string
}
